sound/soc/codecs/rt715-sdca.c

```diff
@@ -88,171 +135,189 @@ static int rt715_sdca_set_amp_gain_put(struct snd_kcontrol *kcontrol,
-	} else {
-		ctl_l = val_l;
-		if (ctl_l & BIT(15)) {
-			ctl_l = ~(val_l - 1) & 0xffff;
-			neg_flag = 1;
-		}
-		ctl_l *= 1000;
-		ctl_l >>= 8;
-		if (neg_flag)
-			ctl_l = mc->shift - ctl_l / RT715_SDCA_DB_STEP;
+	for (i = 0; i < 2; i++) {
+		ret = regmap_read(rt715->mbq_regmap, mc->reg + i, &val);
+		if (ret < 0)
+			dev_err(component->dev, "Failed to read 0x%x, ret=%d\n",
+				mc->reg + i, ret);
 		else
-			ctl_l = mc->shift + ctl_l / RT715_SDCA_DB_STEP;
+			ucontrol->value.integer.value[i] =
+				rt715_sdca_get_gain(val, mc->shift);
 	}
 
-	neg_flag = 0;
-	/* R channel */
-	if (mc->shift == 8) {
-		ctl_r = (val_r >> mc->shift) / 10;
-	} else {
-		ctl_r = val_r;
-		if (ctl_r & BIT(15)) {
-			ctl_r = ~(val_r - 1) & 0xffff;
-			neg_flag = 1;
-		}
-		ctl_r *= 1000;
-		ctl_r >>= 8;
-		if (neg_flag)
-			ctl_r = mc->shift - ctl_r / RT715_SDCA_DB_STEP;
+	return 0;
+}
+
+static int rt715_sdca_set_amp_gain_4ch_get(struct snd_kcontrol *kcontrol,
+		struct snd_ctl_elem_value *ucontrol)
+{
+	struct snd_soc_component *component = snd_kcontrol_chip(kcontrol);
+	struct rt715_sdca_priv *rt715 = snd_soc_component_get_drvdata(component);
+	unsigned int reg_base = kcontrol->private_value, i;
+	const unsigned int gain_sft = 0x2f;
+	unsigned int val;
+	int ret;
+
+	for (i = 0; i < 4; i++) {
+		ret = regmap_read(rt715->mbq_regmap, reg_base + i, &val);
+		if (ret < 0)
+			dev_err(component->dev, "Failed to read 0x%x, ret=%d\n",
+				reg_base + i, ret);
 		else
-			ctl_r = mc->shift + ctl_r / RT715_SDCA_DB_STEP;
+			ucontrol->value.integer.value[i] =
+				rt715_sdca_get_gain(val, gain_sft);
 	}
 
-	ucontrol->value.integer.value[0] = ctl_l;
-	ucontrol->value.integer.value[1] = ctl_r;
-
 	return 0;
 }
 
-static const DECLARE_TLV_DB_SCALE(in_vol_tlv, -17625, 375, 0);
-static const DECLARE_TLV_DB_SCALE(mic_vol_tlv, 0, 1000, 0);
-
-static int rt715_sdca_fu_info(struct snd_kcontrol *kcontrol,
-			 struct snd_ctl_elem_info *uinfo)
+static int rt715_sdca_set_amp_gain_8ch_get(struct snd_kcontrol *kcontrol,
+		struct snd_ctl_elem_value *ucontrol)
 {
-	uinfo->type = SNDRV_CTL_ELEM_TYPE_INTEGER;
-	uinfo->count = 2;
-	uinfo->value.integer.min = 0;
-	uinfo->value.integer.max = 1;
+	struct snd_soc_component *component = snd_kcontrol_chip(kcontrol);
+	struct rt715_sdca_priv *rt715 = snd_soc_component_get_drvdata(component);
+	unsigned int reg_base = kcontrol->private_value;
+	const unsigned int gain_sft = 8;
+	unsigned int val_l, val_r;
+	unsigned int i, reg;
+	int ret;
+
+	for (i = 0; i < 8; i += 2) {
+		ret = regmap_read(rt715->mbq_regmap, reg_base + i, &val_l);
+		if (ret < 0)
+			dev_err(component->dev, "Failed to read 0x%x, ret=%d\n",
+					reg_base + i, ret);
+		else
+			ucontrol->value.integer.value[i] = (val_l >> gain_sft) / 10;
+
+		reg = (i == 6) ? (reg_base - 1) | BIT(15) : reg_base + 1 + i;
+		ret = regmap_read(rt715->mbq_regmap, reg, &val_r);
+		if (ret < 0)
+			dev_err(component->dev, "Failed to read 0x%x, ret=%d\n",
+					reg, ret);
```

Comment:
Should we return ret if ret < 0? The ucontrol->value.integer.value is not valid now.